client/app/cloud/project/compute/infrastructure/virtualMachine/add/cloud-project-compute-infrastructure-virtualMachine-add.service.js

```diff
@@ -0,0 +1,181 @@
+            osType: type
+        }), flavor => this.CloudFlavorService.augmentFlavor(flavor)), {
+            diskType: "ssd",
+            flex: false
+        });
+    }
+
+    getFilteredFlavorsByRegion (flavors, regionCode) {
+        const filteredFlavors = _.uniq(_.remove(flavors, { region: regionCode }), "name");
+        const usedFlavorNames = _.uniq(_.map(filteredFlavors, flavor => flavor.name));
+        const notAvailableFlavors = _.filter(flavors, flavor => !_.include(usedFlavorNames, flavor.name));
+        const outOfRegionFlavors = _.map(_.uniq(notAvailableFlavors, "name"), flavor => {
+            flavor.regions = _.map(_.filter(notAvailableFlavors, f => f.name === flavor.name), "region");
+            flavor.disabled = "NOT_AVAILABLE";
+            delete flavor.region;
+            delete flavor.price;
+            return flavor;
+        });
+
+        return filteredFlavors.concat(outOfRegionFlavors);
+    }
+
+    getFilteredPrivateNetworksByRegion (privateNetworks, regionCode, subNets = []) {
+        return _.chain(privateNetworks)
+            .filter(network => {
+                if (!_.has(subNets, network.id)) {
+                    return false;
+                }
+                return _.some(network.regions, "region", regionCode);
+            })
+            .sortBy("vlanId")
+            .map(network => {
+                const pad = Array(5).join("0");
+                return _.assign(network, {
+                    vlanId: pad.substring(0, pad.length - network.vlanId.toString().length) + network.vlanId
+                });
+            })
+            .value();
+    }
+
+    getImageApps (images) {
+        return _.uniq(_.forEach(this.CloudImageService.getApps(images), app => {
+            app.appName = _.get(app, "name", "")
+                .replace(/^[a-z0-9\s]+\ -\ /ig, "")
+                .replace(" - deprecated", "");
+            delete app.region;
+            delete app.id;
+        }), "name");
+    }
+
+    getPrivateNetworksSubNets (serviceName, privateNetworks) {
+        let networkIds = [];
+        return _.chain(privateNetworks)
+            .map(_.property("id"))
+            .tap(ids => (networkIds = ids))
+            .map(networkId => this.OvhApiCloudProjectNetworkPrivateSubnet.Lexi().query({ serviceName, networkId }).$promise)
+            .thru(promises => { // .mapKeys on a more recent lodash.
+                const collection = {};
+                _.forEach(promises, (promise, key) => {
+                    collection[networkIds[key]] = promise;
+                });
+                return this.$q.all(collection);
+            })
+            .value()
+            .then(subNets => subNets)
+            .catch(() => []);
+    }
+
+    getRegionsByImageType (regions, allImages, imageType) {
+        if (this.CloudImageService.isSnapshot(imageType)) {
+            return _.filter(regions, region => _.get(imageType, "region", "") === _.get(region, "microRegion.code"));
```

Comment:
_.get(region, "microRegion.code") => can microregion be undefined?  Is it normal?  When I look at the function I'm thinking it would be weird to get undefined because microRegion.code was not defined.  Maybe this code should break when this happen.

Same thing goes for imageType => can it be undefined?  When? why?

Reply:
ok, done